esm_parser/yaml_to_dict.py

```diff
@@ -66,16 +79,61 @@ def constructor_env_variables(loader, node):
         :return: the parsed string that contains the value of the environment
         variable
         """
+        # file name (including the full path if not in PWD)
+        fname = node.start_mark.name
+
+        # line number (starts from 0) where match is found
+        line_num = node.start_mark.line
+
+        # start and end of the colums numbers in the line (starts from 0)
+        # For character index from the start of the file use node.start_mark.index
+        column_start = node.start_mark.column
+        column_end = node.end_mark.column
+        
+        # read the file into a list of line. This can also be achieved by 
+        # read() instead but since the YAML files are not large (max few KBs),
+        # this is a quicked solution
+        file_lines = open(fname, 'r').readlines()
+        
+        # current line without the newline at the end 
+        cur_line = file_lines[line_num].rstrip()
+        
+        # Print extra debug messages if ESM_PARSER_DEBUG environment variable is set
+        if os.environ.get("ESM_PARSER_DEBUG", None):
+            print()
+            print("===== esm_parser -> yaml_to_dict.py -> create_env_loader() =====")
+            print(f"::: Reading the file: {fname}")
+            print("::: Reading the line:")
+            print(f">>>{cur_line}<<<")
+            print(f"::: found the match: >>>{cur_line[column_start:column_end]}<<<")
+            print("==================== END OF ESM_PARSER_DEBUG ===================")
+            print()
+        
         value = loader.construct_scalar(node)
-        match = pattern.findall(value)  # to find all env variables in line
-        if match:
-            full_value = value
-            for g in match:
-                full_value = full_value.replace(
-                    f'${{{g}}}', os.environ.get(g, g)
-                )
-                loader.env_variables.append((g, full_value))
-            return full_value
+        
+        # Check if we have a valid !ENV tag on the line
+        envtag_match = re.search(pattern_envtag, cur_line)
+        
+        # list of matches for ${VAR}. Each match is VAR
+        envvar_matches = pattern_envvar.findall(value) 
+
+        # Parse the environmental variables only if the line has a valid !ENV tag
+        if envtag_match:
+            if envvar_matches: 
+                full_value = value
+                for env_var in envvar_matches:
+
+                    # first check if the variable exists in the shell environment
+                    if not os.getenv(env_var): 
+                        print(f"!!! ERROR: {env_var} is not an environment variable. Exiting")
+                        sys.exit(1)
```

Comment:
Not a major thing, but some time ago I put together the `esm_parser.user_error` method. This method does exactly what you do in this two lines, in one single line and allows us to keep some consistency on the way we report errors to the user. 

If you would want to use it here then it will look like something like this:
```
esm_parser.user_error(f"{env_var} is not defined", f"{env_var} is not an environment variable. Exiting")
```

Then the output you'll get would be:
```

ERROR: <env_var> is not defined
-------------------------------

<env_var> is not an environment variable. Exiting

```

Just so you know, a very similar method called `user_note` exists for notifying things to the user, in fact, `user_error` calls `user_note` for the printing and then just adds sys.exit(1).

You can check this methods here:
https://github.com/esm-tools/esm_parser/blob/release/esm_parser/esm_parser.py#L2391-L2421

Reply:
I think in some point not so far away, we want to move away from printing and start using `logging` or `loguru`. Using `user_error` and `user_note` to communicate with the user will save us some time when it comes to changing from `print` to a logging module.